File: blackjack/deck.cpp

#include "deck.h"
#include "card.h"
#include <array>
#include <unordered_map>
#include <iostream>
using namespace std;

// namespace BJ {


Deck::Deck() { }

Deck::~Deck() { }

Deck::Deck(int num_decks) 
{
	total_cards = CARDS_PER_DECK * num_decks;

	deck_map.insert({"A", 4 * num_decks});
	deck_map_name.insert({"A", "ACE"});
	deck_map.insert({"10", 16 * num_decks});
	deck_map_name.insert({"10", "TEN"});
	deck_map.insert({"9", 4 * num_decks});
	deck_map_name.insert({"9", "NINE"});
	deck_map.insert({"8", 4 * num_decks});
	deck_map_name.insert({"8", "EIGHT"});
	deck_map.insert({"7", 4 * num_decks});
	deck_map_name.insert({"7", "SEVEN"});
	deck_map.insert({"6", 4 * num_decks});
	deck_map_name.insert({"6", "SIX"});
	deck_map.insert({"5", 4 * num_decks});
	deck_map_name.insert({"5", "FIVE"});
	deck_map.insert({"4", 4 * num_decks});
	deck_map_name.insert({"4", "FOUR"});
	deck_map.insert({"3", 4 * num_decks});
	deck_map_name.insert({"3", "THREE"});
	deck_map.insert({"2", 4 * num_decks});
	deck_map_name.insert({"2", "TWO"});

	card_counter[TWO] = 4 * num_decks;
	card_counter[THREE] = 4 * num_decks;
	card_counter[FOUR] = 4 * num_decks;
	card_counter[FIVE] = 4 * num_decks;
	card_counter[SIX] = 4 * num_decks;
	card_counter[SEVEN] = 4 * num_decks;
	card_counter[EIGHT] = 4 * num_decks;
	card_counter[NINE] = 4 * num_decks;
	card_counter[TEN] = 16 * num_decks;
	card_counter[ACE] = 4 * num_decks;
}

void Deck::removeCard(Card *card) 
{
	if (card_counter[card->getName()] > 0) 
	{	
		card_counter[card->getName()]--;
		if (deck_map.find(card->card_name) == deck_map.end())
		{
			cout << "Deck could not remove card. " 
				 << "Card " << card->card_name << "not found\n";
			// throw error
		}
		else
		{
			deck_map[card->card_name]--;
			total_cards--;
		}
	}
	//otherwise throw an exception
	else {

	}
}


void Deck::printDeck() 
{
	cout << "********* DECK *********\n";
	cout << "total #cards: " << total_cards << "\n";
	for (auto it = deck_map.begin(); it != deck_map.end(); ++it)
	{
		cout << it->first << ": " << it->second << "\n";
	}
}

	
// } // namespace BJ

File: BlackjackApp/BlackjackLib/Deck.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BlackjackLib
{
    public class Deck
    {
        #region Private Members

        private Dictionary<CardName, int> cardCountDict;
        private int totalCardCount;
        private int deckCount;

        #endregion

        #region Public Members

        public const int MAX_DECK_COUNT = 8;

        public int DeckCount {  get { return deckCount; } }

        public Dictionary<CardName, int> CardCountDict { get { return cardCountDict; } }

        public int TotalCardCount { get { return totalCardCount; } }

        public static Dictionary<CardName, int> CardValDict = new Dictionary<CardName, int>
        {
            { CardName.TWO, 2 },
            { CardName.THREE, 3 },
            { CardName.FOUR, 4 },
            { CardName.FIVE, 5 },
            { CardName.SIX, 6 },
            { CardName.SEVEN, 7 },
            { CardName.EIGHT, 8 },
            { CardName.NINE, 9 },
            { CardName.TEN, 10 },
            { CardName.ACE, 11 }

        };

        public static Dictionary<int, PossibleOutcome> PossibleOutcomeValues = new Dictionary<int, PossibleOutcome>
        {
            { 17, PossibleOutcome.SEVENTEEN },
            { 18, PossibleOutcome.EIGHTEEN },
            { 19, PossibleOutcome.NINETEEN },
            { 20, PossibleOutcome.TWENTY },
            { 21, PossibleOutcome.TWENTYONE }
        };

        #endregion


        #region Constructor

        public Deck() : this(4)
        {
            deckCount = 4;
        }

        public Deck(int numDecks)
        {
            deckCount = numDecks;

            cardCountDict = new Dictionary<CardName, int>
            {
                { CardName.TWO, 4 * numDecks },
                { CardName.THREE, 4 * numDecks },
                { CardName.FOUR, 4 * numDecks },
                { CardName.FIVE, 4 * numDecks },
                { CardName.SIX, 4 * numDecks },
                { CardName.SEVEN, 4 * numDecks },
                { CardName.EIGHT, 4 * numDecks },
                { CardName.NINE, 4 * numDecks },
                { CardName.TEN, 16 * numDecks },
                { CardName.ACE, 4 * numDecks }

            };

            totalCardCount = 52 * numDecks;


        }

        #endregion

        #region Public Functions

        public Deck Clone()
        {
            return new Deck
            {
                cardCountDict = new Dictionary<CardName, int>(this.cardCountDict),
                totalCardCount = this.totalCardCount
            };
        }

        public decimal GetCardProb(CardName cardName)
        {
            if (cardName == CardName.NONE) return 0;
            return (decimal)cardCountDict[cardName] / totalCardCount;
        }

        public void RemoveCard(CardName cardName)
        {
            if (cardCountDict[cardName] <= 0) return;

            cardCountDict[cardName]--;
            totalCardCount--;
        }

        public List<CardName> GetRemainingCards()
        {
            return cardCountDict
                .Where(d => d.Value > 0)
                .Select(d => d.Key)
                .OrderBy(d => d)
                .ToList();
        }

        public void AddCard(CardName cardName)
        {
            cardCountDict[cardName]++;
            totalCardCount++;
        }

        public void UpdateDeckCount(int newNumDecks)
        {
            if (newNumDecks < 0 || newNumDecks > MAX_DECK_COUNT) return;

            int deckCountDiff = newNumDecks - deckCount;

            totalCardCount = 0;

            foreach (CardName cardName in cardCountDict.Keys.ToList())
            {
                int cardsPerDeck = 4;
                if (cardName == CardName.TEN)
                {
                    cardsPerDeck = 16;
                    
                }

                cardCountDict[cardName] = Math.Max(0, cardCountDict[cardName] + (cardsPerDeck * deckCountDiff));
                totalCardCount += cardCountDict[cardName];

            }

            deckCount = newNumDecks;
        }

        public void UpdateCardCount(CardName cardName, int newCount)
        {
            // get the amount of change in the card counts
            int cardCountChange = Math.Abs(newCount - cardCountDict[cardName]);
            // if the new amount is less than the old, the change is negative
            if (newCount < cardCountDict[cardName]) cardCountChange *= -1;

            cardCountDict[cardName] = newCount;
            totalCardCount += cardCountChange;
        }

        public static CardName GetCardName(int cardValue)
        {
            return CardValDict
                .Where(c => c.Value == cardValue)
                .Select(c => c.Key)
                .FirstOrDefault();
        }

        #endregion
    }

    public enum CardName
    {
        NONE,
        TWO,
        THREE,
        FOUR,
        FIVE,
        SIX,
        SEVEN,
        EIGHT,
        NINE,
        TEN,
        ACE
    }

    public enum PossibleOutcome
    {
        SIXTEEN,
        SEVENTEEN,
        EIGHTEEN,
        NINETEEN,
        TWENTY,
        TWENTYONE,
        BUST,
        NOBUST
    }
}
